torch/csrc/distributed/spmd/engine.h

```diff
@@ -0,0 +1,105 @@
+#pragma once
+
+#include <torch/csrc/distributed/spmd/event.h>
+#include <torch/csrc/distributed/spmd/event_handler.h>
+
+#include <unordered_map>
+#include <unordered_set>
+
+namespace torch {
+namespace distributed {
+namespace spmd {
+
+// WIP RootHandler
+// TODO: generate POST_FORWARD event
+class TORCH_API RootHandler : public EventHandler {
+ public:
+  using EventHandler::EventHandler;
+  C10_NODISCARD std::vector<EventSchema> ingressEvents() const override;
+  C10_NODISCARD std::vector<EventSchema> egressEvents() const override;
+  std::vector<std::shared_ptr<Future>> handleEvent(
+      const c10::intrusive_ptr<Event>& /* unused */) override;
+};
+
+// The event-based engine that maintains the event-handling graph and routes
+// events to corresponding handlers properly.
+//
+// During construction, the Engine ctor takes a list of EventHandler instances.
+// Each EventHandler specifies its ingress and egress events. Based on that
+// information, the Engine builds a bipartite graph where the two sets of nodes
```

Comment:
I can redraw the following graph in that way if that helps understanding the implementation. The differences are:

event as node: HandlerNode -> EventNode -> HandlerNode
event as edge: HandlerNode -> EventEdge -> HandlerNode

We will still need an Edge struct for events, and the graph traversal code will need to handle both node and edge instead of just nodes. I am OK with either way.

@SciPioneer proposes a Future-based graph. We can also wait for that design to decide on the graph data structure.